### ISSUE TITLE
[Feature] Readiness Probe in multi-node etcd 

### ISSUE THREAD

**Issue by ishan16696** [gardener/etcd-backup-restore#423]
**Feature (What you would like to be added):**
Currently, [readinessProbe of etcd](https://github.com/gardener/etcd-druid/blob/master/charts/etcd/templates/etcd-statefulset.yaml#L55-L60) is set to an endpoint `/healthz` of [HTTP server](https://github.com/gardener/etcd-backup-restore/blob/322d3c74fcd5ea7be9961ce5e9d7ffed48ecc34e/pkg/server/httpAPI.go#L168) running in a backup sidecar.
This behaviour needed to be updated or improved as readinessProbe of `clustered-etcd`  should depend on whether there is `etcd-leader` present or not then only it should serve the incoming write requests.

**Motivation (Why is this needed?):**

**Approach/Hint to the implement solution (optional):**
 Approaches :
1.  `ETCDCTL_API=3 etcdctl endpoint health --endpoints=${ENDPOINTS} --command-timeout=Xs` 
`etcdctl endpoint health` command performs a GET on the "health" key([source](https://github.com/etcd-io/etcd/blob/main/etcdctl/ctlv3/command/ep_command.go#L126-L133))
    - fails when there is no etcd leader or when Quorum is lost as I think GET request will fail if there is no etcd leader present.

     **Advantage** of this Method  (`etcdctl endpoint health`).
       -  We don't have to worry about [such scenarios](https://github.com/gardener/etcd-druid/issues/147) of causing outage as now snapshotter failure won't fails the readinessProbe of etcd.
       -  If there is no Quorum present, kubelet will mark the `etcd-members` as `NotReady` and they won't able to serve the write as well as read requests.
       
      **Disadvantage** of this Method  (`etcdctl endpoint health`).
      - [Owner check feature](https://github.com/gardener/etcd-backup-restore/blob/322d3c74fcd5ea7be9961ce5e9d7ffed48ecc34e/pkg/server/backuprestoreserver.go#L261-L267) depends on endpoint `/healthz` of [HTTP server](https://github.com/gardener/etcd-backup-restore/blob/322d3c74fcd5ea7be9961ce5e9d7ffed48ecc34e/pkg/server/httpAPI.go#L168) because when Owner check fails it fails the readinessProbe of etcd by setting the [HTTP status to 503](https://github.com/gardener/etcd-backup-restore/blob/322d3c74fcd5ea7be9961ce5e9d7ffed48ecc34e/pkg/server/backuprestoreserver.go#L267) but this Owner check in multi-node scenario is already being discussed [here](https://github.com/gardener/etcd-druid/issues/242).
      -  It completely decouples the snapshotter of backup sidecar and readinessProbe of etcd, backup sidecar won't able to control when to let the traffic come in.
      
  
2.  Use endpoint `/healthz` of [HTTP server](https://github.com/gardener/etcd-backup-restore/blob/322d3c74fcd5ea7be9961ce5e9d7ffed48ecc34e/pkg/server/httpAPI.go#L168) running in backup sidecar with modifications in such a way that whenever `backup-restore leader` is elected it should set `HTTP server status to 200` for itself as well for all  `backup-restore followers` and set the `HTTP server status to 503` when there is no etcd-leader present.
    **Advantage** of this Method  (`/healthz`).
     -  We still have some  coupling of `snapshotter` of backup sidecar and `readinessProbe` of etcd, backup sidecar will able to control when to let the traffic come in for etcd.

    **Disadvantage** of this Method  (`/healthz`).
    - It will takes time to implement as well as to handle edge cases.

    **Future Scope:**
    - Go with method 2 as it give us flexibility to set the `readinessProbe` from backup-sidecar and switch to gRPC instead of  sending REST requests.

**Comment by ishan16696** [comment]
cc @timuthy @abdasgupta @aaronfern @shreyas-s-rao

**Comment by timuthy** [comment]
Thanks for the summary @ishan16696. I have two comments or suggestions in general:

1. We should introduce a `/readyz` endpoint for readiness probes. It can have whatever implementation we think is the best w/o touching the pod template, i.e. we are very open to tweak the checks in the future.

2. In general I prefer your **first** approach because it executes a "real" check (w/o being influenced or redirected to another member) in the member pod which makes the check simple and independent. The disadvantages you mentioned can be eliminated by covering everything relevant behind a `/readyz` endpoint.

**Comment by ishan16696** [comment]
> We should introduce a `/readyz` endpoint for readiness probes

I didn’t get how `/readyz` will be different from `/healthz`.
Do you want to execute `etcdctl endpoint health` from backup-restore and then set `/readyz` accordingly or something else?

**Comment by ishan16696** [comment]
added One more method:  `/health` endpoint of `etcd`
Although it is quite similar to Method 1 `etcdctl endpoint health` but I thought it is worth to mention.

**Comment by timuthy** [comment]
My suggestion is just to hide the readiness checks (whatever they are) behind an endpoint because then we are and will be very flexible about the implementation behind it w/o changing the pod spec. Please lets not write something like `etcdctl endpoint health` directly to pod spec.

That endpoint can be called `/readyz` or we change `/healthz` which is also fine for me.

**Comment by ishan16696** [comment]
> Please lets not write something like etcdctl endpoint health directly to pod spec

yes, that’s why I updated the description of issue with one more approach of `/health` endpoint provided by `etcd`.

> My suggestion is just to hide the readiness checks (whatever they are) behind an endpoint because then we are and will be very flexible about the implementation behind it w/o changing the pod spec

make sense

**Comment by timuthy** [comment]
Discussed with @ishan16696, @shreyas-s-rao, @aaronfern the following plan to proceed:

1. Etcd-Druid configures the following readiness probe in the `StatefulSet`:
```
spec:
  template:
    spec:
    
```

**Comment by timuthy** [comment]
Discussed with @ishan16696, @shreyas-s-rao, @aaronfern the following plan to proceed:

1. Etcd-Druid configures the following readiness probe in the `StatefulSet` **if multi-node is used**:
```
spec:
  template:
    spec:
      containers:
        name: etcd
        readinessProbe:
          httpGet:
            path: /health
            port: 2379
            scheme: HTTP
```

2. Once we support owner checks for the multi-node scenario, we modify the `/healthz` endpoint of the `etcd-backup-restore` sidecar to aggregate:

- Calling etcd's built-in `/health` check.
- Executing owner check.

3. Etcd-Druid reverts `readinessProbe` to use `/healthz` again.

**Comment by ishan16696** [comment]
As discussed in our internal sync that right now we have many unknowns since we don't know the what edge cases might arise in production w.r.t `readinessProbe` in **multi-node etcd** as well as **multi-AZ etcd**  we can go ahead with the above approach(https://github.com/gardener/etcd-backup-restore/issues/423#issuecomment-1025645494) for time being and later we can pick this up with better Knowledge of edge cases, testing scenarios as well as better Knowledge of features requirements.

cc @timuthy @abdasgupta @aaronfern 

/on-hold

**Comment by ishan16696** [comment]
This issue needs to be revisited after https://github.com/gardener/etcd-druid/issues/445 etcd version upgrade.
